apps/OboeTester/app/src/main/cpp/ReverseJniEngine.cpp

```diff
@@ -0,0 +1,127 @@
+    if (mAudioStream) {
+        mAudioStream->requestStop();
+        closeAudioStream();
+    }
+}
+
+void ReverseJniEngine::setBufferSizeInBursts(int bufferSizeInBursts) {
+    if (mAudioStream) {
+        mAudioStream->setBufferSizeInFrames(bufferSizeInBursts * mAudioStream->getFramesPerBurst());
+        LOGI("Buffer size set to %d frames.", mAudioStream->getBufferSizeInFrames());
+    }
+}
+
+void ReverseJniEngine::setupAudioStream() {
+    oboe::AudioStreamBuilder builder;
+    builder.setDirection(oboe::Direction::Output)
+            ->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+            ->setSharingMode(oboe::SharingMode::Exclusive)
+            ->setFormat(oboe::AudioFormat::Float)
+            ->setChannelCount(oboe::ChannelCount::Mono)
+            ->setDataCallback(this);
+
+    oboe::Result result = builder.openStream(mAudioStream);
+    if (result != oboe::Result::OK) {
+        LOGE("Failed to create stream. Error: %s", oboe::convertToText(result));
+    }
+}
+
+void ReverseJniEngine::closeAudioStream() {
+    if (mAudioStream) {
+        mAudioStream->close();
+    }
+}
+
+oboe::DataCallbackResult ReverseJniEngine::onAudioReady(oboe::AudioStream *oboeStream,
+                                                        void *audioData,
+                                                        int32_t numFrames) {
+    JNIEnv *env;
+    // Attach the current thread to the JVM to make JNI calls.
+    int getEnvStat = mJavaVM->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    if (getEnvStat == JNI_EDETACHED) {
+        if (mJavaVM->AttachCurrentThread(&env, nullptr) != 0) {
+            LOGE("Failed to attach current thread to JVM");
+            return oboe::DataCallbackResult::Stop;
+        }
+    }
+
+    int xRunCount = mAudioStream->getXRunCount().value();
+
+    // Create a new Java float array to pass the audio data.
+    jfloatArray audioDataArray = env->NewFloatArray(numFrames * mAudioStream->getChannelCount());
+
+    // Call the Java method via reverse JNI.
+    env->CallVoidMethod(mJavaObject, mOnAudioReadyId, audioDataArray, numFrames, xRunCount);
+
+    if (env->ExceptionCheck()) {
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        LOGE("Exception thrown in Java onAudioReady");
+        env->DeleteLocalRef(audioDataArray);
+        return oboe::DataCallbackResult::Stop;
+    }
+
+    // Copy the data from the Java float array to the Oboe audio buffer.
+    jfloat *arrayElements = env->GetFloatArrayElements(audioDataArray, nullptr);
+    memcpy(audioData, arrayElements, static_cast<unsigned long>(numFrames) * mAudioStream->getChannelCount() * sizeof(float));
+    env->ReleaseFloatArrayElements(audioDataArray, arrayElements, JNI_ABORT);
+    env->DeleteLocalRef(audioDataArray);
+
+    // Detach the thread if we attached it.
+    if (getEnvStat == JNI_EDETACHED) {
+        mJavaVM->DetachCurrentThread();
```

Comment:
Attaching and detaching the thread every callback is expensive and affect the performance a lot.

Reply:
Great suggestion! Done!